### ISSUE TITLE
Compiler Panic

### ISSUE THREAD

**Issue by AljoschaMeyer** [rust-lang/rust#47904]
I somehow managed to get the compiler to panic on trying to compile my code. Sorry for the unhelpful description, but that's all I've got. The code uses `#![feature(closure_to_fn_coercion)]`, that might be relevant.

The codebase is [here (links directly to the commit whose changes made the compiler crash)](https://github.com/AljoschaMeyer/packet-stream-rs/commit/215e8f1476c141a5d01e1a19134f872ac8792377)
Crashing command: `cargo test`

I don't have the time to investigate this right now, but will need to do so at some point. I'll post if I manage to narrow down the cause.

`rustc --version --verbose`:
```
rustc 1.19.0-nightly (28fd1e519 2017-05-27)
binary: rustc
commit-hash: 28fd1e519a3807c632d58d01e73b4948000639ba
commit-date: 2017-05-27
host: x86_64-unknown-linux-gnu
release: 1.19.0-nightly
LLVM version: 4.0
```

## Output for `set -x RUST_BACKTRACE full; set -x RUST_BACKTRACE 1; cargo test --verbose -- --nocapture`, omitting warnings for unused code:

```
       Fresh cfg-if v0.1.2
       Fresh void v1.0.2
       Fresh byteorder v1.2.1
       Fresh regex-syntax v0.3.9
       Fresh winapi-build v0.1.1
       Fresh log v0.4.1
       Fresh unreachable v1.0.0
       Fresh winapi v0.2.8
       Fresh log v0.3.9
       Fresh utf8-ranges v0.1.3
       Fresh regex-syntax v0.4.2
       Fresh futures v0.1.17
       Fresh libc v0.2.35
       Fresh utf8-ranges v1.0.0
       Fresh lazy_static v1.0.0
       Fresh memchr v2.0.1
       Fresh memchr v0.1.11
       Fresh rand v0.3.20
       Fresh kernel32-sys v0.2.2
       Fresh aho-corasick v0.6.4
       Fresh aho-corasick v0.5.3
       Fresh iovec v0.1.1
       Fresh thread_local v0.3.5
       Fresh ordermap v0.3.4
       Fresh thread-id v2.0.0
       Fresh bytes v0.4.6
       Fresh regex v0.2.5
       Fresh multi-consumer-stream v0.2.0
       Fresh thread_local v0.2.7
       Fresh tokio-io v0.1.4
       Fresh env_logger v0.4.3
       Fresh multi-producer-sink v0.1.0
       Fresh regex v0.1.80
       Fresh async-ringbuffer v0.1.0
       Fresh quickcheck v0.5.0
       Fresh atm-io-utils v0.1.0
       Fresh env_logger v0.3.5
       Fresh atm-async-utils v0.1.1
       Fresh quickcheck v0.4.1
       Fresh partial-io v0.2.5
   Compiling packet-stream v0.1.0 (file:///home/aljoscha/projects/packet-stream-rs)
     Running `rustc --crate-name packet_stream src/lib.rs --emit=dep-info,link -C debuginfo=2 --test -C metadata=066442e80d647294 -C extra-filename=-066442e80d647294 --out-dir /home/aljoscha/projects/packet-stream-rs/target/debug/deps -L dependency=/home/aljoscha/projects/packet-stream-rs/target/debug/deps --extern multi_consumer_stream=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libmulti_consumer_stream-5a4ce9a776d6c4b8.rlib --extern atm_io_utils=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libatm_io_utils-071ee257d54bd20c.rlib --extern quickcheck=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libquickcheck-97686dbf695e8694.rlib --extern futures=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libfutures-fb03a1e35425c824.rlib --extern tokio_io=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libtokio_io-5be3a1d45098ff0e.rlib --extern rand=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/librand-395d79dcece8a755.rlib --extern async_ringbuffer=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libasync_ringbuffer-3d231e8e522c1ba9.rlib --extern multi_producer_sink=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libmulti_producer_sink-5622659b55680890.rlib --extern partial_io=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libpartial_io-f62cfe2fd8a90c72.rlib --extern atm_async_utils=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libatm_async_utils-79e420050927c989.rlib --extern void=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libvoid-f4e2164e54e8e4e8.rlib`

[unused code warnings omitted]

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'assertion failed: !value.needs_subst()', /checkout/src/librustc/infer/mod.rs:629
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:355
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:365
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:549
   5: std::panicking::begin_panic
   6: rustc::infer::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'tcx>>::erase_late_bound_regions_and_normalize
   7: rustc_trans::monomorphize::resolve_closure
   8: <rustc_trans::collector::MirNeighborCollector<'a, 'tcx> as rustc::mir::visit::Visitor<'tcx>>::visit_rvalue
   9: rustc::mir::visit::Visitor::visit_mir
  10: rustc_trans::collector::collect_items_rec
  11: rustc_trans::collector::collect_items_rec
  12: rustc_trans::collector::collect_items_rec
  13: rustc_trans::collector::collect_items_rec
  14: rustc_trans::collector::collect_items_rec
  15: rustc_trans::collector::collect_items_rec
  16: rustc_trans::collector::collect_items_rec
  17: rustc_trans::collector::collect_items_rec
  18: rustc_trans::collector::collect_items_rec
  19: rustc_trans::collector::collect_items_rec
  20: rustc_trans::collector::collect_items_rec
  21: rustc_trans::collector::collect_items_rec
  22: rustc_trans::collector::collect_items_rec
  23: rustc_trans::collector::collect_items_rec
  24: rustc_trans::collector::collect_items_rec
  25: rustc_trans::collector::collect_items_rec
  26: rustc_trans::base::collect_and_partition_translation_items::{{closure}}
  27: rustc_trans::base::trans_crate
  28: rustc_driver::driver::phase_4_translate_to_llvm
  29: rustc_driver::driver::compile_input::{{closure}}
  30: rustc_driver::driver::phase_3_run_analysis_passes::{{closure}}
  31: rustc_driver::driver::phase_3_run_analysis_passes
  32: rustc_driver::driver::compile_input
  33: rustc_driver::run_compiler

error: Could not compile `packet-stream`.

Caused by:
  process didn't exit successfully: `rustc --crate-name packet_stream src/lib.rs --emit=dep-info,link -C debuginfo=2 --test -C metadata=066442e80d647294 -C extra-filename=-066442e80d647294 --out-dir /home/aljoscha/projects/packet-stream-rs/target/debug/deps -L dependency=/home/aljoscha/projects/packet-stream-rs/target/debug/deps --extern multi_consumer_stream=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libmulti_consumer_stream-5a4ce9a776d6c4b8.rlib --extern atm_io_utils=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libatm_io_utils-071ee257d54bd20c.rlib --extern quickcheck=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libquickcheck-97686dbf695e8694.rlib --extern futures=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libfutures-fb03a1e35425c824.rlib --extern tokio_io=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libtokio_io-5be3a1d45098ff0e.rlib --extern rand=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/librand-395d79dcece8a755.rlib --extern async_ringbuffer=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libasync_ringbuffer-3d231e8e522c1ba9.rlib --extern multi_producer_sink=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libmulti_producer_sink-5622659b55680890.rlib --extern partial_io=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libpartial_io-f62cfe2fd8a90c72.rlib --extern atm_async_utils=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libatm_async_utils-79e420050927c989.rlib --extern void=/home/aljoscha/projects/packet-stream-rs/target/debug/deps/libvoid-f4e2164e54e8e4e8.rlib` (exit code: 101)

```

**Comment by Sunspar** [comment]
There may be a valid reason, but you're using a fairly old nightly build. Specifically, it appears as though `closure_to_fn_coercion` landed in [1.19](https://blog.rust-lang.org/2017/07/20/Rust-1.19.html).

It appears to compile and run tests (the only failing test I noticed was `ps::tests::requests`) on my MBP (running OS X) with:

```
rustc 1.25.0-nightly (def3269a7 2018-01-30)
binary: rustc
commit-hash: def3269a71be2e737cad27418a3dad9f5bd6cd32
commit-date: 2018-01-30
host: x86_64-apple-darwin
release: 1.25.0-nightly
LLVM version: 4.0
```

Are you able to update to a newer nightly compiler?

**Comment by AljoschaMeyer** [comment]
Huh, I'm so used to my package manager keeping things up to date, I forgot I used rustup and need to manually update. Updated and everything works fine again.

Sorry for bothering you with this, and thanks for the quick response!
